kdc/fast.c

```diff
@@ -560,6 +568,83 @@ fast_unwrap_request(astgs_request_t r,
 	free_AP_REQ(&ap_req);
 	if (ret)
 	    goto out;
+
+	ret = krb5_unparse_name(r->context, armor_server,
+				&armor_server_principal_name);
+	if (ret)
+	    goto out;
+
+	ret = _kdc_verify_flags(r->context, r->config,
+				&ticket->ticket,
+				armor_server_principal_name);
+	free(armor_server_principal_name);
+	if (ret) {
+	    _kdc_audit_addreason((kdc_request_t)r,
+				 "Armor TGT expired or invalid");
+	    goto out;
+	}
+
+	if (tgs_ticket != NULL)
+	    flags = HDB_F_FOR_TGS_REQ;
+	else
+	    flags = HDB_F_FOR_AS_REQ;
+	if (r->req.req_body.kdc_options.canonicalize)
+	    flags |= HDB_F_CANON;
+
+	/* Fetch the name from the TGT. */
+	ret = _krb5_principalname2krb5_principal(r->context,
+						 &fast_principal,
+						 ticket->ticket.cname,
+						 ticket->ticket.crealm);
+	if (ret)
+	    goto out;
+
+	/* Look up the name given in the TGT in the database. */
+	ret = _kdc_db_fetch(r->context, r->config, fast_principal,
+			    HDB_F_GET_CLIENT | flags,
+			    NULL, NULL, &fast_client);
+	if (ret) {
+	    ret = KRB5KDC_ERR_C_PRINCIPAL_UNKNOWN;
+	    goto out;
+	}
+
+	/*
+	 * If the account is present in the database, check the account
+	 * flags.
+	 */
+	ret = kdc_check_flags(r,
+			      tgs_ac == NULL,
+			      fast_client,
+			      NULL);
+	if (ret) {
+	    _kdc_free_ent(r->context, fast_client);
+	    goto out;
+	}
+
+	/* Verify the PAC of the TGT. */
+	ret = _kdc_check_PAC(r->context, r->config, fast_principal, NULL,
+			     fast_client, armor_user,
+			     armor_user, armor_user,
+			     &armor_key->key, &r->ticket_key->key,
+			     &ticket->ticket, &ad_kdc_issued, &mspac);
+	_kdc_free_ent(r->context, fast_client);
+	if (ret) {
+	    const char *msg = krb5_get_error_message(r->context, ret);
+
+	    kdc_log(r->context, r->config, 4,
+		    "Verify PAC failed for %s (%s) from %s with %s",
+		    r->sname, r->cname, r->from, msg);
+	    krb5_free_error_message(r->context, msg);
+	    goto out;
+	}
+	if (mspac == NULL || !ad_kdc_issued) {
+	    ret = KRB5KDC_ERR_BADOPTION;
+	    kdc_log(r->context, r->config, 0,
+		    "Armor ticket not signed with PAC; FAST failed (%s).",
+		    mspac ? "Ticket unsigned" : "No PAC");
```

Comment:
This may create a backwards compatibility issue for us because older Heimdal KDCs won't issue PACs unconditionally. I think you will need to hide this behind a feature flag.

Reply:
See #892 for the appropriate flag.